### PR TITLE
[SPARK-20245][SQL][minor] pass output to LogicalRelation directly

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Currently `LogicalRelation` has a `expectedOutputAttributes` parameter, which makes it hard to reason about what the actual output is. Like other leaf nodes, `LogicalRelation` should also take `output` as a parameter, to simplify the logic

## How was this patch tested?

existing tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @gatorsmile

File: sql/core/src/test/scala/org/apache/spark/sql/sources/PathOptionSuite.scala [hunk 1]
```diff
@@ -75,13 +75,13 @@ class PathOptionSuite extends DataSourceTest with SharedSQLContext {
            |USING ${classOf[TestOptionsSource].getCanonicalName}
            |OPTIONS (PATH '/tmp/path')
         """.stripMargin)
-      assert(getPathOption("src") == Some("file:/tmp/path"))
+      assert(getPathOption("src").map(makeQualifiedPath) == Some(makeQualifiedPath("/tmp/path")))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/PathOptionSuite.scala]: These changes in this test suite are not related to this PR, just to reinforce it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala [hunk 2]
```diff
@@ -18,39 +18,21 @@ package org.apache.spark.sql.execution.datasources
...
-  override val output: Seq[AttributeReference] = {
-    val attrs = relation.schema.toAttributes
-    expectedOutputAttributes.map { expectedAttrs =>
-      assert(expectedAttrs.length == attrs.length)
-      attrs.zip(expectedAttrs).map {
-        // We should respect the attribute names provided by base relation and only use the
-        // exprId in `expectedOutputAttributes`.
-        // The reason is that, some relations(like parquet) will reconcile attribute names to
-        // workaround case insensitivity issue.
-        case (attr, expected) => attr.withExprId(expected.exprId)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala]: It sounds like this logics mentioned in the comments is removed by this PR.

**Comment by gatorsmile** [comment]
LGTM except only one comment

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala]: good catch! I found this logic is only useful when converting hive tables to data source tables, so I moved the logic there: https://github.com/apache/spark/pull/17552/files#diff-ee66e11b56c21364760a5ed2b783f863R215

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala]: Agree.

**Comment by gatorsmile** [comment]
LGTM pending Jenkins.

**Comment by cloud-fan** [comment]
thanks for the review, merging to master!
